test/Microsoft.AspNetCore.Sockets.Tests/TestWebSocketConnectionFeature.cs

```diff
@@ -107,23 +108,45 @@ public override void Dispose()
             {
                 try
                 {
-                    await _input.WaitToReadAsync();
-
-                    if (_input.TryRead(out var message))
+                    if (_internalBuffer.Buffer == null || _internalBuffer.Buffer.Length == 0)
                     {
-                        if (message.MessageType == WebSocketMessageType.Close)
+                        await _input.WaitToReadAsync();
+
+                        WebSocketMessage message;
+                        if (_input.TryRead(out message))
                         {
-                            _state = WebSocketState.CloseReceived;
-                            _closeStatus = message.CloseStatus;
-                            _closeStatusDescription = message.CloseStatusDescription;
-                            return new WebSocketReceiveResult(0, WebSocketMessageType.Close, true, message.CloseStatus, message.CloseStatusDescription);
+                            if (message.MessageType == WebSocketMessageType.Close)
+                            {
+                                _state = WebSocketState.CloseReceived;
+                                _closeStatus = message.CloseStatus;
+                                _closeStatusDescription = message.CloseStatusDescription;
+                                return new WebSocketReceiveResult(0, WebSocketMessageType.Close, true, message.CloseStatus, message.CloseStatusDescription);
+                            }
+
+                            _internalBuffer = message;
                         }
+                    }
 
-                        // REVIEW: This assumes the buffer passed in is > the buffer received
-                        Buffer.BlockCopy(message.Buffer, 0, buffer.Array, buffer.Offset, message.Buffer.Length);
+                    var length = _internalBuffer.Buffer.Length;
```

Comment:
You might be able to clean up this code a little bit if you use `Memory<byte>` instead of an array.